### PR TITLE
User can add asset model number

### PR DESCRIPTION
**Pull request by Awinja-Andela**
## What does this PR do?
User can add asset model number 
## Description of Task to be completed?
- Add add button to All Asset Page
- create common modal
- create toast Message action and reducer
- create common dropdown
- create common button
## How should this be manually tested?
N/A

## What are the relevant pivotal tracker stories?
[#157340067](https://www.pivotaltracker.com/story/show/157340067)

## Related PRs branch | PR (branch_name) | (pr_link)
Branch | PR
------ | ------
other_pr_production | [link]()
other_pr_master | [link]()
## Todos
- [x] Raise PR
- [x] Test
- [ ] Documentation
## Screenshots
screen shot 2018-06-12 at 22 13 14

screen shot 2018-06-20 at 15 23 10

### CODE REVIEW

File: src/components/ModalComponent.jsx [hunk 1]
```diff
@@ -0,0 +1,30 @@
+import React from 'react';
+import { Modal, Button, Icon } from 'semantic-ui-react';
+
+
+const AddAssetModel = () => (
+    <Modal trigger={<i className="plus link icon" />} centered={true}>
+    <Modal.Header>Add Model</Modal.Header>
+    <Modal.Content>
+        <div class="ui form">
```
Comment (theterminalguy) [inline src/components/ModalComponent.jsx]: @Awinja-Andela please indent your markups properly.

File: src/components/AssetsComponent.jsx [hunk 2]
```diff
@@ -50,7 +50,9 @@ export class AssetsComponent extends Component {
               <Table.HeaderCell>Sub-category</Table.HeaderCell>
               <Table.HeaderCell>Type</Table.HeaderCell>
               <Table.HeaderCell>Make</Table.HeaderCell>
-              <Table.HeaderCell>Model</Table.HeaderCell>
+              <Table.HeaderCell className="header">Model
```
Comment (theterminalguy) [inline src/components/AssetsComponent.jsx]: @Awinja-Andela please indent your markups properly.

**Comment by theterminalguy** [comment]
@Awinja-Andela please indent your markups properly.

**Comment by Awinja-Andela** [comment]
@HawiCaesar tests are failing.

**Comment by Awinja-Andela** [comment]
@sachungo review?

File: src/components/AssetsComponent.jsx [hunk 3]
```diff
@@ -48,19 +54,55 @@ export class AssetsComponent extends Component {
...
           <Table.Header>
             <Table.Row>
-              <Table.HeaderCell>Category</Table.HeaderCell>
-              <Table.HeaderCell>Sub-category</Table.HeaderCell>
-              <Table.HeaderCell>Type</Table.HeaderCell>
-              <Table.HeaderCell>Make</Table.HeaderCell>
-              <Table.HeaderCell>Model</Table.HeaderCell>
-              <Table.HeaderCell>Item</Table.HeaderCell>
+              <Table.HeaderCell>
+                  <div className="header">
```
Comment (sachungo) [inline src/components/AssetsComponent.jsx]: The indentation is off from here to end of `render` function. Please fix and ensure you are using indentation of `2 spaces`

File: src/components/AssetsComponent.jsx [hunk 4]
```diff
@@ -48,19 +54,55 @@ export class AssetsComponent extends Component {
...
+                  </div>
+              </Table.HeaderCell>
+              <Table.HeaderCell>
+                  <div className="header">
+                  Sub-category
+                  <ModalComponent/>
+                </div>
+              </Table.HeaderCell>
+              <Table.HeaderCell >
+                  <div className="header">
```
[the remark this answers is not in the document]
Comment (sachungo) [reply src/components/AssetsComponent.jsx]: @HawiCaesar Wasn't this turned off?

File: src/components/AssetsComponent.jsx [hunk 5]
```diff
@@ -86,9 +128,10 @@ export class AssetsComponent extends Component {
           </Table.Footer>
         </Table>
       </SideMenuComponent>
-    )
-  }
-};
+      );
+  } // render()
+
```
Comment (sachungo) [inline src/components/AssetsComponent.jsx]: No need for the extra line

File: src/components/AssetsComponent.jsx [hunk 6]
```diff
@@ -86,9 +128,10 @@ export class AssetsComponent extends Component {
           </Table.Footer>
         </Table>
       </SideMenuComponent>
-    )
-  }
-};
+      );
+  } // render()
```
Comment (sachungo) [inline src/components/AssetsComponent.jsx]: No need for the comment

File: src/components/AssetsComponent.jsx [hunk 7]
```diff
@@ -86,9 +128,10 @@ export class AssetsComponent extends Component {
...
+
+}; // AssetComponent
```
Comment (sachungo) [inline src/components/AssetsComponent.jsx]: Comment is unnecessary

File: src/components/CardComponent.jsx [hunk 8]
```diff
@@ -1,4 +1,4 @@
-import React, { Component } from 'react'
+import React from 'react'
```
[the remark this answers is not in the document]
Comment (sachungo) [reply src/components/CardComponent.jsx]: @HawiCaesar Didn't you turn off `import/extensions`?

File: src/components/PieChartComponent.jsx [hunk 9]
```diff
@@ -1,7 +1,4 @@
 import React from 'react';
-import ReactDOM from 'react-dom';
-import FusionCharts from 'fusioncharts';
-import Charts from 'fusioncharts/fusioncharts.charts';
```
Comment (sachungo) [inline src/components/PieChartComponent.jsx]: Are all the deletions in this file unused imports?

File: src/components/common/ButtonComponent.jsx [hunk 10]
```diff
@@ -0,0 +1,32 @@
+import React from 'react';
+import { Button } from 'semantic-ui-react';
+import PropTypes from 'prop-types';
+
+const ButtonComponent = (props) => {
+  const { buttonName, color, onClick } = props;
```
Comment (sachungo) [inline src/components/common/ButtonComponent.jsx]: You can also destructure the props from the param as:
```js
const ButtonComponent = ({ buttonName, color, onClick }) => {...}

File: src/components/common/ModalComponent.jsx [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+      )
+    );
+    return (
+      <div>
+        <Modal
+          trigger={
+            <i
+              className="plus link icon"
+              onClick={this.toggleModal}
+              onKeyUp={() => {}}
```
Comment (sachungo) [inline src/components/common/ModalComponent.jsx]: Is this handler necessary when it does nothing?

File: src/components/common/ModalComponent.jsx [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+export default class ArtModal extends Component {
+  state = { modalOpen: false }
+
+  toggleModal = () => this.setState({ modalOpen: !this.state.modalOpen });
+
+  render() {
+    const { children } = this.props;
+    const childrenWithProps = React.Children.map(children, child =>
+      React.cloneElement(child,
```
[the remark this answers is not in the document]
Comment (sachungo) [reply src/components/common/ModalComponent.jsx]: Ignore this houndCI comment for now whenever you see it

File: src/components/common/ModalComponent.jsx [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+      </div>
+    );
+  }
+}
+ArtModal.propTypes = {
+  children: PropTypes.node.isRequired,
+  modalTitle: PropTypes.string.isRequired
+};
+
+ArtModal.defaultProps = {
```
Comment (sachungo) [inline src/components/common/ModalComponent.jsx]: As houndCI says, you don't need a defaultProp for props with `isRequired`. You either remove the `isRequired` or delete the `defaultProps`

File: src/components/common/TextInputComponent.jsx [hunk 14]
```diff
@@ -0,0 +1,18 @@
+import React from 'react';
+import { Input } from 'semantic-ui-react';
+import PropTypes from 'prop-types';
```
[the remark this answers is not in the document]
Comment (sachungo) [reply src/components/common/TextInputComponent.jsx]: If `linebreak-style` is still a problem, then add `"linebreak-style": 0` to `.eslintrc`

File: src/_constants/index.js [hunk 15]
```diff
@@ -14,6 +14,9 @@ const CONSTANTS = KeyMirror({
   CREATE_ASSET_FAIL: true,
   LOAD_FEEDBACK_SUCCESS: true,
   LOAD_FEEDBACK_FAILURE: true,
+  CREATE_MODEL_NUMBER_SUCCESS: true,
+  CREATE_MODEL_NUMBER_FAILURE: true,
```
Comment (sachungo) [inline src/_constants/index.js]: Seems `prettier` isn't working properly. It should have removed this trailing comma. I'm going to look more into it so for the meantime, remove the trailing comma and empty line 19

File: src/components/common/ModalComponent.jsx [hunk 16]
```diff
@@ -0,0 +1,55 @@
+import React, { Component } from 'react';
+import { Modal } from 'semantic-ui-react';
```
[the remark this answers is not in the document]
Comment (sachungo) [reply src/components/common/ModalComponent.jsx]: @HawiCaesar Didn't you turn off import/extensions?

File: src/_constants/index.js (hunk 15 above)
Comment (sachungo) [reply src/_constants/index.js]: Hmm..if it doesn't do its work then we can remove it and just use `eslint` and `stylelint` for now. There's no need to spend so much time on a chore task, when eslint is already working. Please do not remove the precommit hook as it will still be used with the other linters
 
@HawiCaesar can you again make a quick PR that does the following:
1. removes prettier
2. updates `"react/forbid-prop-types": ['error', { forbid: ['any', 'object'] }]`  to `"react/forbid-prop-types": ["error", { "forbid": ["any", "object"] }]` in `.eslintrc`. Note the use of single quote in the first version
3. Disables `import/extensions`
4. Disables `function-paren-newline`

This will take about 10 minutes to do

File: src/components/common/ModalComponent.jsx (hunk 11 above)
Comment (HawiCaesar) [reply src/components/common/ModalComponent.jsx]: This is necessary to close the modal

File: src/components/AssetsComponent.jsx (hunk 7 above)
Comment (HawiCaesar) [reply src/components/AssetsComponent.jsx]: Noted. I will change this.

File: src/_constants/index.js (hunk 15 above)
Comment (HawiCaesar) [reply src/_constants/index.js]: @sachungo Alright. Working on it.

File: src/components/PieChartComponent.jsx (hunk 9 above)
Comment (HawiCaesar) [reply src/components/PieChartComponent.jsx]: They were not being used when the dashboard loads up. The terminal would flag them as `no-unused-vars`

File: src/_constants/index.js (hunk 15 above)
Comment (HawiCaesar) [reply src/_constants/index.js]: @sachungo If I remove `prettier` that means this precommit
```json
"*.{js,jsx,json,css}": [
        "./node_modules/.bin/prettier --write",
        "git add"
      ]
```
will not work and will give out an error. Shouldn't I remove the above precommit hook instead ? Please clarify ?

Comment (sachungo) [reply src/_constants/index.js]: Apologies for the ambiguity. I meant the eslint and stylelint precommit hooks. The prettier one is to be deleted

**Comment by sachungo** [comment]
@HawiCaesar @Awinja-Andela Please let me know when the conflict and feedback have been addressed

File: src/_actions/assetMakes.actions.js [hunk 17]
```diff
@@ -26,6 +21,6 @@ export const loadAssetMakes = () => {
  *
  * @return {object} type and payload
  */
-export const loadAssetMakesSuccess = (assetMakes) => {
-  return { type: LOAD_ASSET_MAKES_SUCCESS, payload: assetMakes };
-}
+export const loadAssetMakesSuccess = assetMakes => (
+  { type: LOAD_ASSET_MAKES_SUCCESS, payload: assetMakes }
```
Comment (sachungo) [inline src/_actions/assetMakes.actions.js]: This is a comment that can be ignored for now, when using `()` as a return for arrow function, it's aesthetically pleasing to structure the contents as given in the example
```
export const loadAssetMakesSuccess = assetMakes => ({
  type: LOAD_ASSET_MAKES_SUCCESS,
  payload: assetMakes
});
```
This should be the same for all similar instances in the current changes 🙂

File: src/_actions/modelNumbers.actions.js [hunk 18]
```diff
@@ -28,6 +29,41 @@ export const loadModelNumbers = () => {
...
-  return { type: LOAD_ASSET_MODEL_NUMBERS, payload: modelNumbers };
-}
+export const loadModelNumbersSuccess = modelNumbers => (
+  { type: LOAD_ASSET_MODEL_NUMBERS, payload: modelNumbers }
+);
+
+/**
+ * create new asset model number thunk
+ * @param {object} newModel details of new model to be created
+ *
```
Comment (sachungo) [inline src/_actions/modelNumbers.actions.js]: @HawiCaesar this line hasn't been deleted yet

File: src/_actions/modelNumbers.actions.js [hunk 19]
```diff
@@ -28,6 +29,41 @@ export const loadModelNumbers = () => {
...
+*/
```
Comment (sachungo) [inline src/_actions/modelNumbers.actions.js]: This is off...move forward by 1 space. Check the indentation of line 31 for a demonstration

File: src/_components/ModelNumber/ModelNumberContainer.jsx [hunk 20]
```diff
@@ -0,0 +1,115 @@
...
+
+import ModelNumberComponent from '../../components/ModelNumber/ModelNumberComponent';
+import { ToastMessage } from '../../_utils/ToastMessage';
+
+import { loadAssetMakes } from '../../_actions/assetMakes.actions';
+import { createModelNumbers } from '../../_actions/modelNumbers.actions';
+import resetToastMessageContent from '../../_actions/resetToastMessage.actions';
+
+class ModelNumberContainer extends React.Component {
+  constructor(props) {
```
Comment (sachungo) [inline src/_components/ModelNumber/ModelNumberContainer.jsx]: Use of the constructor is unnecessary if there's no state initialized with props. You can declare state as a class property like so:
```
class ModelNumberContainer extends React.component {
  state = {
    modelNumber: '',
    assetMake: '',
    saveButtonState: false,
    ...
  }

  static getDerivedStateFromProps(nextProps, prevState) {...}
```

File: src/_components/RoutesComponent.jsx [hunk 21]
```diff
@@ -2,11 +2,11 @@ import React, { Component } from 'react';
 import { Route, Switch } from 'react-router-dom';
 
 import AuthenticateComponent from './AuthenticateComponent';
-import AssetTypes from '../components/AssetTypesComponent';
+import AssetTypesComponent from '../components/AssetTypesComponent';
```
Comment (sachungo) [inline src/_components/RoutesComponent.jsx]: I believe changes to this file are unnecessary. Please revert to what is in develop

File: src/_css/ModalComponent.css [hunk 22]
```diff
@@ -0,0 +1,4 @@
+.ui.fullscreen.modal {
+    width: 100%!important;
+
```
Comment (sachungo) [inline src/_css/ModalComponent.css]: Delete the empty space

File: src/_css/AssetComponent.css [hunk 23]
```diff
@@ -0,0 +1,4 @@
+.header{
+    display: flex;
```
[the remark this answers is not in the document]
Comment (sachungo) [reply src/_css/AssetComponent.css]: @HawiCaesar I'm still seeing 4 spaces for the css files 🤷‍♀️

File: src/components/AssetsComponent.jsx [hunk 24]
```diff
@@ -1,112 +1,145 @@
...
+import TableRowComponent from './TableRowComponent';
+import ModalComponent from './common/ModalComponent';
+import SideMenuComponent from '../_components/SideMenuComponent';
+import { getAssetsAction } from '../_actions/assets.action';
+import { createModelNumbers } from '../_actions/modelNumbers.actions';
+import ModelNumberContainer from '../_components/ModelNumber/ModelNumberContainer';
+import '../_css/AssetComponent.css';
+
+export class AssetsComponent extends Component {
+  state = {
```
Comment (sachungo) [inline src/components/AssetsComponent.jsx]: This is a good example of _initializing state as a class variable_ I was talking about in the previous comment

File: src/components/AssetsComponent.jsx [hunk 25]
```diff
@@ -1,112 +1,145 @@
...
+                <div className="header">
+                  Item
+                  <ModalComponent />
+                </div>
+              </Table.HeaderCell>
+            </Table.Row>
+          </Table.Header>
+
+          <Table.Body>
+            {
```
Comment (sachungo) [inline src/components/AssetsComponent.jsx]: Why not `{this.loadTableContent()} `? Anyway, it's not that important

File: src/components/AssetsComponent.jsx [hunk 26]
```diff
@@ -1,112 +1,145 @@
...
+
+const mapStateToProps = ({ viewAssets }) => {
+  const { assets, assetsCount } = viewAssets;
+  return {
+    assets,
+    assetsCount,
+  };
+};
+
+export default connect(mapStateToProps, {
```
Comment (sachungo) [inline src/components/AssetsComponent.jsx]: Nice 👍

File: src/components/BarlineGraphComponent.jsx [hunk 27]
```diff
@@ -1,208 +1,210 @@
...
-  render() {
-    return (
-      <div className="barlineGraph">
-        <Divider hidden />
-        <ReactFC {...this.chartConfigs} />
-      </div>
-    );
-  }
-}
+import React from 'react';
```
Comment (sachungo) [inline src/components/BarlineGraphComponent.jsx]: Please revert the changes to this file to what is in develop as your task doesn't have anything to do with it 🙂

File: src/components/CardComponent.jsx [hunk 28]
```diff
@@ -1,20 +1,18 @@
-import React, { Component } from 'react';
+import React from 'react';
 import { Card } from 'semantic-ui-react';
 import PropTypes from 'prop-types';
 
-class CardComponent extends Component {
+class CardComponent extends React.Component {
```
Comment (sachungo) [inline src/components/CardComponent.jsx]: Why the change to this line? It is unnecessary.  Revert the import to `import React, { Component } from 'react';` and the class definition to `class CardComponent extends Component`

File: src/components/CardComponent.jsx [hunk 29]
```diff
@@ -24,10 +22,10 @@ class CardComponent extends Component {
 }
 
 CardComponent.propTypes = {
-  statistics: PropTypes.arrayOf(PropTypes.object)
+  statistics: PropTypes.array
 };
 
-CardComponent.defaultProps = {
+CardComponent.defauktTypes = {
```
Comment (sachungo) [inline src/components/CardComponent.jsx]: Typo. Please revert to previous code of `CardComponent.defaultProps`

File: src/_components/ModelNumber/ModelNumberContainer.jsx [hunk 30]
```diff
@@ -0,0 +1,115 @@
...
+    super(props);
+    this.state = {
+      modelNumber: '',
+      assetMake: '',
+      saveButtonState: false,
+      toastMessageContent: {
```
[the remark this answers is not in the document]
Comment (Awinja-Andela) [reply src/_components/ModelNumber/ModelNumberContainer.jsx]: @sachungo this state is being used.i am unsure why hound is flagging it.

File: src/_components/RoutesComponent.jsx [hunk 31]
```diff
@@ -23,8 +23,8 @@ class RoutesComponent extends Component {
           <AuthenticateComponent
             exact
             isAuthenticated={this.checkAuthentication()}
-            path="/asset_types"
-            component={AssetTypes}
+            path='/asset_types'
```
[the remark this answers is not in the document]
Comment (Awinja-Andela) [reply src/_components/RoutesComponent.jsx]: @sachungo too scared to touch this violation, because this is how the file is in develop.

File: src/_components/RoutesComponent.jsx [hunk 32]
```diff
@@ -35,8 +35,8 @@ class RoutesComponent extends Component {
           <AuthenticateComponent
             exact
             isAuthenticated={this.checkAuthentication()}
-            path="/feedback"
-            component={Feedback}
+            path='/feedback'
```
[the remark this answers is not in the document]
Comment (Awinja-Andela) [reply src/_components/RoutesComponent.jsx]: @sachungo too scared to touch this violation, because this is how the file is in develop.

**Comment by Awinja-Andela** [comment]
@sachungo @houndci-bot is speechless :-D
